core/slice.go

```diff
@@ -121,7 +121,7 @@ func (sl *Slice) Append(header *types.Header, domPendingHeader *types.Header, do
 	location := header.Location()
 
 	// Don't append the block which already exists in the database.
-	if sl.hc.HasHeader(header.Hash(), header.NumberU64()) {
+	if sl.hc.HasHeader(header.Hash(), header.NumberU64()) && (sl.hc.GetTd(header.Hash(), header.NumberU64()) != nil) {
```

Comment:
What is the scenario where you would have a header, but not the TD? Is this just a debug check that needs to go away?

Reply:
No, this is put purposefully. This is done during the skeletonHeader construction. During the skeleton fetch, we need to know if we actually appended and not just we have the data, because fetcher might have given us the data which gives no guarantees that we have data between the skeleton, or someone might have broadcasted that data and we don't have anything around it.

Reply:
For any block that is not appended we don't have the TD, so we need this check.